app/models/tag_set.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class TagSet < ApplicationRecord
+  belongs_to :tag_group, class_name: 'TagGroup', optional: false
+  belongs_to :tag2_group, class_name: 'TagGroup', optional: true
+
+  # We can assume adapter_type is the same for both tag groups
+  # So we delegate to either group
```

Comment:
Given that TagSet is allowed to be a single index tag set, having the second tag group (tag2_group) null, the comment is not quite right. We delegate to the first only to cover this case.

Although the validation when TagSet is dual index tag set ensures the adapter types are equal, we still delegate to the first because of the case above.

Reply:
Good point, ill update the comment